### PR TITLE
[WIP] Folder structure

### LINKED ISSUE #22: Allow for different years in different repos
The current default setup appears to be:

```
- advent
  - src
    - year1
    - year2
  - data
    - year1
    - year2
```

I much prefer

```
- advent
  - year1
    - src
    - data
  - year2
    - src
    - data
```

Are you willing to make this structure optional? Simplest would be to just have an option to omit the year from the path entirely, and leave the higher structure to the user. I'd be happy to make a PR

### PR DESCRIPTION
**Pull request by kescobo**
This will close #22 

I also added an `include_year` kwarg (default `true`) to the relevant functions. Setting it to `false` will leave the year off of the path.

I managed to fix most of the tests locally, but there are some failures that I'm not sure about, possibly related to the `Mocking` package which I'm passingly familiar with, but which I've never used. It's also possible that it's something obvious that I'm overlooking, but from what I can tell, it's trying to download data from 2020 day 23, but that's obviously not available yet.

### CODE REVIEW

File: src/AdventOfCode.jl [hunk 1]
```diff
@@ -21,13 +21,16 @@ function _download_data(year, day)
     error("Unable to download data")
 end
 
-function _template(year, day)
-    data_path = normpath(pwd() * "data/$year/day_$day.txt")
+function _template(year, day; include_year = true)
+    rel_path = "data/day_$day.txt"
+    include_year && (rel_path = joinpath(string(year), rel_path))
+    
+    data_path = normpath(pwd() * rel_path) # this isn't being used?
```
Comment (SebRollen) [inline src/AdventOfCode.jl]: You're right, this isn't used, we can remove it

File: src/AdventOfCode.jl [hunk 2]
```diff
@@ -21,13 +21,16 @@ function _download_data(year, day)
...
     """
     # $(_base_url(year, day))
     using AdventOfCode
 
-    input = readlines("data/$year/day_$day.txt")
+    input = readlines("$rel_path")
```
Comment (SebRollen) [inline src/AdventOfCode.jl:32]: we can probably skip the string interpolation here:
```suggestion
    input = readlines(rel_path)
```

**Comment by SebRollen** [comment]
> there are some failures that I'm not sure about

That was me having written some tests that only pass in december 😅 I've removed that test now. Left some comments for you above, apart from those this looks good to go!

File: src/AdventOfCode.jl (hunk 2 above)
Comment (kescobo) [reply src/AdventOfCode.jl:32]: I tried that first actually, but that gets expanded to eg `readlines(2020/data/day2.txt)`, which doesn't parse 🙃

**Comment by kescobo** [comment]
Cool! I'll try to wrap it up tomorrow.

File: src/AdventOfCode.jl (hunk 2 above)
Comment (SebRollen) [reply src/AdventOfCode.jl:32]: Ah that's right, I missed that this was inside the triple-quoted block. The current code looks good to me now

File: test/runtests.jl [hunk 3]
```diff
@@ -70,14 +70,14 @@ Mocking.activate()
...
+                    @test isdir("2019/data/")
+                    @test isfile("2019/data/day_1.txt")
+                    @test readlines("2019/data/day_1.txt") == ["TESTDATA"]
+                    @test isdir("2019/src")
+                    @test isfile("2019/src/day_1.jl")
+                    @test String(read("2019/src/day_1.jl")) == AdventOfCode._template(2019, 1)
+                    @test_logs (:warn, r"2019/data/day_1.txt already exists.") (:warn, r"2019/src/day_1.jl already exists.") AdventOfCode.setup_files(2019, 1, force = false)
                 finally
                     rm("data", recursive = true, force = true)
                     rm("src", recursive = true, force = true)
```
Comment (SebRollen) [inline test/runtests.jl]:
```suggestion
                    rm("2019", recursive = true, force = true)
```

File: test/runtests.jl [hunk 4]
```diff
@@ -55,9 +55,9 @@ Mocking.activate()
...
                 AdventOfCode._setup_data_file(2019, 1)
-                @test isdir("data/2019")
-                @test isfile("data/2019/day_1.txt")
-                @test readlines("data/2019/day_1.txt") == ["TESTDATA"]
+                @test isdir("2019/data/")
+                @test isfile("2019/data/day_1.txt")
+                @test readlines("2019/data/day_1.txt") == ["TESTDATA"]
                 @test_logs (:warn, r"will not redownload") AdventOfCode._setup_data_file(2019, 1)
             finally
                 rm("data", recursive = true, force = true)
```
Comment (SebRollen) [inline test/runtests.jl]: We need to change this line too now to ensure that the test-folders are cleaned up
```suggestion
                rm("2019", recursive = true, force = true)
```

**Comment by kescobo** [comment]
So, in trying to add tests for the `include_year = false` situation, I went down a bit of a rabbit hole and did a bunch of reorganization of the tests. But all of the old ones are there (minus the ones you deleted), and I've added the `false` case. They're in a loop now, which adds a bit of abstraction, but I think it's clearer. Let me know if you disagree and I can reset and separate them out

**Comment by kescobo** [comment]
The tests for the `false` case did catch a couple of bonehead errors I made though :laughing:

**Comment by kescobo** [comment]
Lol... hold on, apparently I screwed up the clean-up code, because it deleted the test folder :facepalm:

**Comment by kescobo** [comment]
Failures are on windows, because of path differences in the template output

```
"# https://adventofcode.com/2019/day/1\nusing AdventOfCode\n\ninput = readlines(\"2019/data/day_1.txt\")\n\nfunction part_1(input)\n    nothing\nend\n@info part_1(input)\n\nfunction part_2(input)\n    nothing\nend\n@info part_2(input)\n"
```
vs
```
"# https://adventofcode.com/2019/day/1\nusing AdventOfCode\n\ninput = readlines(\"2019\\data\\day_1.txt\")\n\nfunction part_1(input)\n    nothing\nend\n@info part_1(input)\n\nfunction part_2(input)\n    nothing\nend\n@info part_2(input)\n"
```

Unfortunately, I don't have time to work on this today, but will try to fix up during the holiday.

**Comment by SebRollen** [comment]
@kescobo I managed to solve the Windows errors on version 1.3 in [this branch](https://github.com/SebRollen/AdventOfCode.jl/pull/25/files/bacdd63f61cb00fa868543425bc498ca6e926a11..ae7b27f23d62d2d51c302b5e8f76904c84fcd129). I can't push to your branch directly, but if you copy over the changes from that branch into here, I think we're ready to merge. The nightly version on windows is still failing, but I'll tackle that in another PR

**Comment by kescobo** [comment]
Sorry - busy weekend. Planning to do this first thing tomorrow.

**Comment by kescobo** [comment]
Alright, I think that's good. 

Incidently, I think you _could_ have pushed to this branch, though to be honest, I'm not sure exactly how that works (I think it's just a matter of branch naming)

image

**Comment by SebRollen** [comment]
LGTM! Thanks for adding this feature, will bump the version and tag a release :)

**Comment by kescobo** [comment]
My pleasure! Thanks for the help getting through it :-). Looking forward to starting to use it tomorrow!

**Pull request merged**